Build debs with pbuilder

shanemcd opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

Debian packaging
##### ANSIBLE VERSION

N/A
##### SUMMARY

We started seeing failures related to build dependencies when building 16.04 debs on our 14.04 Jenkins node. `pbuilder` ensures a clean-room environment inside a chroot.

shanemcd committed:
Build debs with pbuilder

shanemcd committed:
Update README in packaging/debian

jlaska commented:
Thanks for updating the README's too, this is a welcome change to use `pbuilder`!

bcoca commented:
So i would not eliminate the old way of building it, just add a new one. some users are stuck in older versions of debian and/or do not use pbuilder. It is easy enough to have this 'in addition to'

jlaska commented:
@bcoca normally I would agree, however the old way of building the DEBs produces DEBs _only_ for the platform you are building on.  In addition, there was logic in there that resulted in failed builds depending on the state of packages installed on the system.  Chroot builds is much preferred to provide consistent builds and accurately test `BuildRequires`.

shanemcd committed:
Add Dockerfile for building debs

ghjm commented:
`DEB_DIST='precise trusty xenial' make deb` worked on my Ubuntu 14.04 box with pbuilder and build deps installed. Looks good to me.

shanemcd committed:
Add local_deb makefile target

Allows users to build debs using locally installed dependencies. This was the `deb` target before moving to pbuilder.

shanemcd commented:
@bcoca Great suggestion. I've added a `local_deb` target to allow for building outside of a chroot.

mattclay commented on packaging/debian/README.md:
````diff
@@ -3,26 +3,35 @@ Ansible Debian Package
+```
+apt-get install cdbs debootstrap devscripts make pbuilder python-setuptools
+git clone git://github.com/ansible/ansible.git
+cd ansible
````
Add `git submodule update --init` here.

mattclay commented on packaging/debian/README.md:
````diff
@@ -3,26 +3,35 @@ Ansible Debian Package
-    sudo apt-get install python-support
+```
+git clone git://github.com/ansible/ansible.git
+cd ansible
````
Add `git submodule update --init` here.

mattclay commented on packaging/debian/README.md:
````diff
@@ -3,26 +3,35 @@ Ansible Debian Package
-    cd ansible
-    make deb
+```
+apt-get install cdbs debootstrap devscripts make pbuilder python-setuptools
````
Additional packages needed on a fresh install of Ubuntu 16.04:
- git
- asciidoc

mattclay commented on packaging/debian/README.md:
```diff
@@ -3,26 +3,35 @@ Ansible Debian Package
 
 To create an Ansible DEB package:
```
It should be noted that running as root is required.

mattclay reviewed (changes requested):
I've successfully tested this using the provided instructions for docker.

However, I'm unable to produce a deb package running in a fresh Ubuntu 16.04 VM. The build process finishes without errors, but no deb package is produced.

mattclay commented on packaging/debian/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM ubuntu:xenial
+
+RUN apt-get update && apt-get install -y \
+    cdbs \
```
The `asciidoc` package is needed.

shanemcd commented:
@mattclay Thanks for the review. I've implemented your feedback, and am currently debugging on a fresh 16.04 VM to diagnose:

> The build process finishes without errors, but no deb package is produced.

For some reason, it is placing the deb in a top-level directory at `/deb-build` when running with sudo, but works when running directly as root. It works fine when running on a 14.04 box. ¯_(ツ)_/¯ Will report back when I find something.

shanemcd commented:
@mattclay I've pushed a fix that works on a fresh 16.04 box. It works when using GNU Make's `CURDIR` variable instead of `PWD`. Also tested with the Docker based approach.

mattclay commented on packaging/debian/README.md:
```diff
@@ -3,26 +3,39 @@ Ansible Debian Package
-    git clone git://github.com/ansible/ansible.git
-    cd ansible
-    make deb
+__Note__: You must either run as root or have write access to `/var/cache/pbuilder`
```
@shanemcd Write access to `/var/cache/pbuilder` isn't enough. The build process fails with this error when not running as root:

```
E: debootstrap can only run as root
```

bcoca commented:
IIRC pbuilder should be doing fakeroot by default
